libraries/AP_HAL/UARTDriver.h

```diff
@@ -160,4 +170,57 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
 
     // disable TX/RX pins for unusued uart
     virtual void disable_rxtx(void) const {}
+
+#if AP_UART_MONITOR_ENABLED
+    // a way to monitor all reads from the UART, putting them in a
+    // buffer. Used by AP_Periph for GPS debug
+    bool set_monitor_read_buffer(ByteBuffer *buffer) {
+        _monitor_read_buffer = buffer;
+        return true;
+    }
+#endif
+
+protected:
+    // key for a locked port
+    uint32_t lock_write_key;
+    uint32_t lock_read_key;
+
+    /*
+      backend begin method
+     */
+    virtual void _begin(uint32_t baud, uint16_t rxSpace, uint16_t txSpace) = 0;
+
+    /*
+      backend write method
+     */
+    virtual size_t _write(const uint8_t *buffer, size_t size) = 0;
+
+    /*
+      backend read method
+     */
+    virtual ssize_t _read(uint8_t *buffer, size_t count)  WARN_IF_UNUSED = 0;
+
+    /*
+      end control of the port, freeing buffers
+     */
+    virtual void _end() = 0;
+
+    /*
+      flush any pending data
+     */
+    virtual void _flush() = 0;
+
+    // check available data on the port
+    virtual uint32_t _available() = 0;
+
+    // discard incoming data on the port
+    virtual bool _discard_input(void) { return false; };
```

Comment:
```suggestion
    virtual bool _discard_input(void) = 0;
```

Reply:
changed - I didn't do it this was as some of the backends didn't implement _discard_input(), I've now tracked down the missing ones and added them